staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_test.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	runtimetesting "k8s.io/apimachinery/pkg/runtime/testing"
+	"k8s.io/apimachinery/pkg/test"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/component-base/version"
+)
+
+func TestEmulatedStorageVersion(t *testing.T) {
+	cases := []struct {
+		name             string
+		scheme           *runtime.Scheme
+		binaryVersion    schema.GroupVersion
+		effectiveVersion version.EffectiveVersion
+		want             schema.GroupVersion
+	}{
+		{
+			name:             "pick compatible",
+			scheme:           GetTestScheme(),
+			binaryVersion:    v1beta1,
+			effectiveVersion: version.NewEffectiveVersion("1.32"),
+			want:             v1alpha1,
+		},
+		{
+			name:             "alpha has been replaced, pick binary version",
+			scheme:           GetTestSchemeWithReplacedAlpha(),
+			binaryVersion:    v1beta1,
+			effectiveVersion: version.NewEffectiveVersion("1.32"),
+			want:             v1beta1,
+		},
+	}
```

Comment:
nit: Might make sense to add a test for the multiple-replacements/replacement-chain case.  Something like:
 
- CronJobV1Alpha1 (replaced by v1alpha2)
- CronJobV1Alpha2 (replaced by v1beta1)
- CronJob v1beta1 (final version)

Reply:
> nit: Might make sense to add a test for the multiple-replacements/replacement-chain case. Something like:
> 
> * CronJobV1Alpha1 (replaced by v1alpha2)
> * CronJobV1Alpha2 (replaced by v1beta1)
> * CronJob v1Beta1 (final version)

I think we do need a test suite for this functionality. I'll admit I only added the minimum tests needed to show that this PR fixed a particular problem and didn't break the adjacent case that should still work.  I think we need a separate PR that fully fleshes the test coverage here.